### DISCUSSION TITLE
MSP+Displayport packets

### DISCUSSION THREAD

**Discussion by r1000ru** [ExpressLRS/ExpressLRS#2574] [Ideas]
Many users use pwm receivers without a flight controller, but with a digital video system. It would be great to implement a uart msp protocol that would:
1. listen to requests for a status package and give it back, including the arm state
2. in the background it would send displayport packets about signal quality, rssi dbm, voltage, current (if sensors are connected), bar altitude (if a barometer is connected)

**Comment by rizangg** [comment]
Hi everyone — following up on this thread since it's exactly the use case I built for.

I run a PWM receiver as the vehicle controller on an RC ground rover (no flight controller), with a DJI O3 Air Unit for FPV video. Based on the receiver-side DisplayPort work that came out of this discussion, I built a rover-focused fork that renders a lightweight OSD directly from the receiver: craft name, arm/standby state, FAILSAFE, steering %, gas %, RSSI, LQ, and (as of the latest build) battery voltage with a configurable low-battery warning.

Repo: https://github.com/rizangg/RZGX-ELRS-Rover

A few things that might be relevant to a possible upstream path:

Tested and field-validated on two different PWM targets (BETAFPV PWM 2.4GHz RX and RadioMaster ER5A/C V2), with repeated indoor/outdoor runs, obstructed-range tests out to ~150 m, and 2S/4S battery voltage validation.
FAILSAFE is enforced as the highest-priority OSD message over any other state (arming, low-battery warning), and that priority order was explicitly tested.
Release notes and SHA-256 checksums for every promoted build, plus a small CI check that verifies binaries against their manifests and catches cross-target string leakage.

Everything here is currently scoped as rover/PWM-specific (steering + gas instead of roll/pitch/throttle), so it's a fairly opinionated fork rather than a drop-in generic feature. But if there's interest in folding any of this back upstream — even just the low-battery-warning state machine or the rover OSD layout as an alternate mode — happy to help extract a minimal patch. Also want to credit @aldyduino here, whose earlier ESP32MSPDisplayPort work was my starting reference.
